File: 042_DeleteNodeInABST.py
"""
Topic: Binary Search Tree
Problem: Delete Node in a BST
Description: Delete node with given key and return new root.
Time: O(h) where h is tree height
Space: O(h) recursion
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def deleteNode(root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
    if not root:
        return None
    if key < root.val:
        root.left = deleteNode(root.left, key)
    elif key > root.val:
        root.right = deleteNode(root.right, key)
    else:
        # found node
        if not root.left:
            return root.right
        if not root.right:
            return root.left
        # two children: replace with inorder successor
        succ = root.right
        while succ.left:
            succ = succ.left
        root.val = succ.val
        root.right = deleteNode(root.right, succ.val)
    return root


def build(vals):
    if not vals: return None
    nodes = [TreeNode(v) if v is not None else None for v in vals]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root


def inorder(root):
    return inorder(root.left) + [root.val] + inorder(root.right) if root else []


if __name__ == '__main__':
    root = build([5,3,6,2,4,None,7])
    root = deleteNode(root, 3)
    print(inorder(root))  # [2,4,5,6,7]

    root = build([5,3,6,2,4,None,7])
    root = deleteNode(root, 5)
    print(inorder(root))  # [2,3,4,6,7]

File: 042_DeleteNodeInABST.cpp
// Topic: BST / Recursion
// Delete Node in a BST: Remove a node with given key and return new root.
// Time Complexity: O(h)
// Space Complexity: O(h)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* findMin(TreeNode* node)
{
    while (node->left) node = node->left;
    return node;
}

TreeNode* deleteNode(TreeNode* root, int key)
{
    if (!root)
        return nullptr;
    if (key < root->val)
        root->left = deleteNode(root->left, key);
    else if (key > root->val)
        root->right = deleteNode(root->right, key);
    else
    {
        if (!root->left)
        {
            TreeNode* r = root->right;
            delete root;
            return r;
        }
        if (!root->right)
        {
            TreeNode* l = root->left;
            delete root;
            return l;
        }
        TreeNode* succ = findMin(root->right);
        root->val = succ->val;
        root->right = deleteNode(root->right, succ->val);
    }
    return root;
}

void inorder(TreeNode* r)
{
    if (!r)
        return;
    inorder(r->left);
    cout << r->val << ' ';
    inorder(r->right);
}

int main()
{
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(2);
    root->left->right = new TreeNode(4);
    root->right->right = new TreeNode(7);
    inorder(root);
    cout << "\n";
    root = deleteNode(root, 3);
    inorder(root);
    cout << "\n";
    root = deleteNode(root, 5);
    inorder(root);
    cout << "\n";
    return 0;
}
